### PR TITLE
🎉 Source Jira: Support all available entities in Jira Cloud

### LINKED ISSUE #1389: Support all available entities in JIRA cloud
## Tell us about the problem you're trying to solve
Support the following streams in Jira on a full refresh basis: 

Application roles
Audit records
Avatars
Dashboards
Filters
Filter sharing
Group and user picker
Groups
Issues
Issue attachments
Issue comments
Issue comment properties
Issue fields
Issue field configurations
Issue custom field contexts
Issue custom field options
Issue custom field options (apps)
Issue links
Issue link types
Issue navigator settings
Issue notification schemes
Issue priorities
Issue properties
Issue remote links
Issue resolutions
Issue search
Issue security level
Issue security schemes
Issue types
Issue type schemes
Issue type screen schemes
Issue type properties
Issue votes
Issue watchers
Issue worklogs
Issue worklog properties
Jira expressions
Jira settings
JQL
Labels
Myself
Permissions
Permission schemes
Projects
Project avatars
Project categories
Project components
Project email
Project key and name validation
Project permission schemes
Project properties
Project roles
Project role actors
Project types
Project versions
Screens
Screen tabs
Screen tab fields
Screen schemes
Server info
Tasks
Time tracking
Users
User properties
User search
Webhooks
Workflows
Workflow transition rules
Workflow schemes
Workflow scheme project associations
Workflow scheme drafts
Workflow statuses
Workflow status categories
Workflow transition properties
App properties
Dynamic modules

FYI we already support some of these streams. 

Please make sure that:

1. In the standard tests for this source, every stream is tested (and that every stream outputs records) unless there is a rate limiting issue
1. you create fake data for every stream -- whoever is working on this issue, feel free to reach out to me if you need support

This is a humungous API, so you are encourages to break up this task into a few tickets e.g: first support all `Workflow`-related streams, then all `Project`-related streams, etc..

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368079337048) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by vitaliizazmic**
## What
Implementing support of streams from issue #1389

## How
* Vetting Jira API documentation and filtering streams, which can be implemented
* Creating schemes for available streams
* Implementing full refresh syncing
* Updating configured catalogs and docs

## Pre-merge Checklist
* [ ] _Run integration tests_
* [ ] _Publish Docker images_

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368146448194) by [Unito](https://www.unito.io)

### CODE REVIEW

File: airbyte-integrations/connectors/source-jira/source_jira/client.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/client.py:94]: We'll probably want to move away from this method to support very large scales, but this is not necessary right now. Currently this is caching all ticket IDs in memory for re-use (which is efficient from a query standpoint but not from a space standpoint) but for a large organization (say 50+ engineering teams, each with a JIRA project with 1000+ issues) this might start to get large. I think this could carry us a long way though. 

On a separate note, I think this is exposing a potentially big problem with the `BaseClient` abstraction which is: it assumes each stream will be synced in isolation. This is not true for child streams like `issue_comments` or `issue_properties`. The better way to sync those would be to have something like: 

```
if is_selected(parent_stream) or is_selected(parent_stream.child_streams):
  for record in parent_stream:
   if is_selected(parent_stream): 
      yield record

   for child in parent_stream.selected_child_streams: 
       yield from list(child)
```

This way we iterate over the parent stream exactly once and in a streaming manner so we don't have to keep in memory anything that is not strictly necessary.

File: airbyte-integrations/connectors/source-jira/source_jira/entities.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/entities.py:31]: it's confusing that some entities don't have the `url` property set. It's also not good from a DRY point of view because this object is supposed to be the source of truth for the URL. Could we instead always include `url` in this object (or at least as much as we can), and in the consuming context use `url.format(params)`?

File: airbyte-integrations/connectors/source-jira/source_jira/client.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/client.py:188]: it seems we are always calling `_get_issue_keys` followed by `item.get("key")`. Why not DRY this inside `_get_issue_keys` thereby also saving space?

File: airbyte-integrations/connectors/source-jira/source_jira/entities.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/entities.py:46]: shouldn't this be extracting the `value` key?

File: airbyte-integrations/connectors/source-jira/source_jira/client.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/source_jira/client.py:169]: isn't this extracting the `keys` property and not `value` property? this is a bit confusing. I don't think this is emitting the correct schema.

**Review by sherifnada — CHANGES_REQUESTED** [review]
Very exciting! 🎉  great to have broader support for the JIRA connector. Though I think we need a bit more work before we are ready for releasing this upgrade. Some comments: 

I think this PR also exposes a big problem with `BaseClient`. We don't have to fix it in this PR, but the way it is currently structured makes it hard to sync child streams efficiently. 

### Testing 
Have we populated these streams with data so we can test that this code functions correctly? Ideally we populate each of these streams with 50+ objects (we'll probably want to do this from the API). This is a big task, so here is my suggestion for how to do it: 
1. Once you iterate on the feedback in this ticket, we consider it "code complete" but blocked on the tickets we'll create in step 2
2. Create two issues for populating JIRA with data. The first github issue should be for all issue-related and project-related streams. The second issue should be for the remaining streams. 
3. We'll prioritize those issues in this sprint or next sprint, according to available bandwidth on GA and Freshdesk. 
4. To resolve those issues, we should create two (or more) PRs which contain test fixture code to create the resources via the API, like the file you merged in this PR to create 50 projects. 
5. Once those issues are resolved, we should merge this into this PR, run tests on this PR, then merge it

I realize this is a pretty big increase in scope, but we really shouldn't be releasing connectors unless we are _very_ confident that they work well. Otherwise it will provide a really bad UX for users if things break that we could have caught otherwise.

File: airbyte-integrations/connectors/source-jira/source_jira/entities.py
Comment (vitaliizazmic) [reply airbyte-integrations/connectors/source-jira/source_jira/entities.py]: Yes, it isn't possible to test this two streams now. JQL response contains list, which items order always different. And Server info returns server time, which obviously different in in each response. testIdenticalFullRefreshes fail for both of them. I disabled them for now and forgot exclude from here.

File: airbyte-integrations/connectors/source-jira/source_jira/client.py
Comment (vitaliizazmic) [reply airbyte-integrations/connectors/source-jira/source_jira/client.py:59]: There are some streams related with another streams. E.g, Project permission schemes. Sometimes such entities does not exist and API return 404 error. I think, in this case we should return empty list of items.

File: airbyte-integrations/connectors/source-jira/source_jira/client.py
Comment (vitaliizazmic) [reply airbyte-integrations/connectors/source-jira/source_jira/client.py:70]: _enumerate_methods is responsible for mapping stream with method for fetching data for it from API. Initially data for  all streams was got by calling lists method with passing relevant parameters. But in the case with related streams logic is more complicated. It is needed to do calling lists method in cycle. Therefore I created custom methods for this streams. This change is for handling both cases - initial and custom.

File: airbyte-integrations/connectors/source-jira/source_jira/client.py
Comment (vitaliizazmic) [reply airbyte-integrations/connectors/source-jira/source_jira/client.py:83]: "Resolutions" it is "Issued resolutions", I renamed it for clarifying.

File: airbyte-integrations/connectors/source-jira/source_jira/client.py
Comment (vitaliizazmic) [reply airbyte-integrations/connectors/source-jira/source_jira/client.py:94]: It is good solution.

File: airbyte-integrations/connectors/source-jira/source_jira/client.py
Comment (vitaliizazmic) [reply airbyte-integrations/connectors/source-jira/source_jira/client.py]: In case of Issue properties stream the first step would be taking property [keys](https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-issue-properties/#api-rest-api-3-issue-issueidorkey-properties-get) and then directly properties.

File: airbyte-integrations/connectors/source-jira/source_jira/client.py
Comment (sherifnada) [reply airbyte-integrations/connectors/source-jira/source_jira/client.py:70]: What's an example of a stream that can't have a `stream__` method in this class and therefore has to be added via this method?

**Review by sherifnada — APPROVED** [review]
LGTM so far. Can you create the tickets as described in the previous review and we will prioritize them into this or next sprint?

File: airbyte-integrations/connectors/source-jira/source_jira/client.py
Comment (sherifnada) [reply airbyte-integrations/connectors/source-jira/source_jira/client.py:70]: Oh I think I understand -- it's the other way around: some methods are very simple and are just a call to `lists`. But more complicated methods must have a more specific implementation. I see. Created the following suggestion to add this context into the code

```suggestion
        # Many streams are just a wrapper around a call to lists() with some preconfigured params. 
        # However, more complicated streams require a custom implementation, which is expressed via 
        # a stream__XYZ method. The latter streams are captured via super's _enumerate_methods(),
        # and the former streams are added here via a partial over lists().
        mapping = super(Client, self)._enumerate_methods()
```

**Comment by sherifnada** [comment]
blocked on #2100 and #2101

File: airbyte-integrations/connectors/source-jira/acceptance-test-config.yml
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jira/acceptance-test-config.yml:15]: can we trim down the catalog and test it on a subset of the streams?

**Review by sherifnada — APPROVED** [review]
@vitaliizazmic LGTM  with one comment on acceptance test catalog💪🏼  think you can release it soon? Almost there 🎉

File: airbyte-integrations/connectors/source-jira/acceptance-test-config.yml
Comment (vitaliizazmic) [reply airbyte-integrations/connectors/source-jira/acceptance-test-config.yml:15]: Yes, I will try to do it.

**Review by sherifnada — APPROVED** [review]

**Pull request merged**
